akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowPrefixAndDownstreamSpec.scala

```diff
@@ -0,0 +1,219 @@
+        .run
+
+      val ex = suffixF.failed.futureValue
+      ex.getCause should not be null
+      (ex.getCause should have).message("I hate mondays! (10)")
+    }
+
+    "propagate flow failures" in assertAllStagesStopped {
+      val (prefixF, suffixF) = Source(0 until 100)
+        .prefixAndDownstreamMat(10) { prefix =>
+          Flow[Int].mapMaterializedValue(_ => prefix).map {
+            case 15 => sys.error("don't like 15 either!")
+            case n  => n
+          }
+        }(Keep.right)
+        .toMat(Sink.ignore)(Keep.both)
+        .run
+      prefixF.futureValue should ===(0 until 10)
+      val ex = suffixF.failed.futureValue
+      //ex.printStackTrace()
+      (ex should have).message("don't like 15 either!")
+    }
+
+    "produce multiple elements per input" in assertAllStagesStopped {
+      val (prefixF, suffixF) = src10()
+        .prefixAndDownstreamMat(7) { prefix =>
+          Flow[Int].mapMaterializedValue(_ => prefix).mapConcat(n => List.fill(n - 6)(n))
+        }(Keep.right)
+        .toMat(Sink.seq[Int])(Keep.both)
+        .run()
+
+      prefixF.futureValue should ===(0 until 7)
+      suffixF.futureValue should ===(7 :: 8 :: 8 :: 9 :: 9 :: 9 :: Nil)
+    }
+
+    "succeed when upstream produces no elements" in assertAllStagesStopped {
+      val (prefixF, suffixF) = Source
+        .empty[Int]
+        .prefixAndDownstreamMat(7) { prefix =>
+          Flow[Int].mapMaterializedValue(_ => prefix).mapConcat(n => List.fill(n - 6)(n))
+        }(Keep.right)
+        .toMat(Sink.seq[Int])(Keep.both)
+        .run()
+
+      prefixF.futureValue should be(empty)
+      suffixF.futureValue should be(empty)
+    }
+
+    "apply materialized flow's semantics when upstream produces no elements" in assertAllStagesStopped {
+      val (prefixF, suffixF) = Source
+        .empty[Int]
+        .prefixAndDownstreamMat(7) { prefix =>
+          Flow[Int].mapMaterializedValue(_ => prefix).mapConcat(n => List.fill(n - 6)(n)).prepend(Source(100 to 101))
+        }(Keep.right)
+        .toMat(Sink.seq[Int])(Keep.both)
+        .run()
+
+      prefixF.futureValue should be(empty)
+      suffixF.futureValue should ===(100 :: 101 :: Nil)
+    }
+
+    "handles upstream cancellation" in assertAllStagesStopped {
+      val publisher = TestPublisher.manualProbe[Int]()
+      val subscriber = TestSubscriber.manualProbe[Int]()
+
+      val flow = Flow[Int].prefixAndDownstreamMat(2) { prefix =>
+        println("materializing flow")
+        Flow[Int].mapMaterializedValue(_ => prefix).prepend(Source(100 to 101)).alsoTo(Sink.foreach(println(_)))
+      }(Keep.right)
+
+      val flow_ = Flow[Int]
+        .alsoToMat {
+          Flow[Int].take(2).toMat(Sink.seq)(Keep.right)
+        }(Keep.right)
+        .drop(2)
+        .prepend(Source(100 to 101))
+        .alsoTo(Sink.foreach(println(_)))
+
+      val matValue = Source
+        .fromPublisher(publisher)
+        .viaMat(if (true) flow else flow_)(Keep.right)
```

Comment:
?

Reply:
debugging leftovers, tried figuring out if the problem was in the test or test-subject code.